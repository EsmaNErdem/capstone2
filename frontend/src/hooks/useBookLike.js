import { useState, useEffect, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user book like
 * This hook provides functionality to manage user likes on books, including retrieving initial like status and count,
 * updating the like status and count, and handling user interactions to like or unlike a book.
 * 
 * -Utilizes useContext data to send user like to database
 * -Creates state to update book like count and user book like status
 * -Use useEffect to avoid to recalculations
 * 
 * Returns user like status, book like count states and handleLike function
 */
const useBookLike = (bookId, initialLikes, bookData={}) => {
    console.debug("useBookLike"); 

    const { hasLikedBook, likeBook } = useContext(UserContext);
    const [liked, setLiked] = useState();
    const [likes, setLikes] = useState();
    const [error, setError] = useState(null);

    // Helper function to convert "undefined" or "null" strings to actual undefined
    const convertUndefined = (value) => (value === "undefined" || value === null ? undefined : value);

    // Convert undefined/null strings in the bookData object
    const convertedBookData = Object.fromEntries(
        Object.keys(bookData).map((key) => [key, convertUndefined(bookData[key])])
    );
        
    /**By using the useEffect, the liked status is only recalculated when the id avoiding unnecessary recalculations on every render.  */
    useEffect(() => {
      setLiked(hasLikedBook(bookId));
      setLikes(initialLikes)
    }, [bookId, initialLikes]);
    
    /**
     * Send user like to database
     * Updates the like count and status based on user actions.
     */ 
    const handleLikeBook = async () => {
        try {
            setError(null);
            if (liked) {
                const unlikedBookId = await likeBook(bookId);
                if (unlikedBookId === bookId) {
                    setLikes((likes) => likes - 1);
                    setLiked(false);
                }
            } else {
                const likedBookId = await likeBook(bookId, convertedBookData);
                if (likedBookId === bookId) {
                    setLikes((likes) => likes + 1);
                    setLiked(true);
                }
            }
        } catch (error) {
            setError('Error liking book.');
            console.error('Error handling book like:', error);
        }
    };
    
  return { liked, likes, error, handleLikeBook };
};

export default useBookLike;