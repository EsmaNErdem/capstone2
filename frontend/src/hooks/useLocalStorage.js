import { useState, useEffect } from "react";

/**
 * Custom Hook to save user info like token and username into localStorage
 * 
 * Creates a user state and looks in localStorage for current user info, if cannot find it, assign null
 * When user info updates, useEffect either cleans it or updates localStorage
 * 
 * Retunrs [user, setUser]
 */

const useLocalStorage = (localStorageKey, initialValue = null) => {
    let initialUserValue
    if(localStorage.getItem(localStorageKey)) {
        try {
            initialUserValue = JSON.parse(localStorage.getItem(localStorageKey))
        } catch (e) {
            console.error("useLocalStorage custom hook: problem loading localStorage data", e);
        }
    } else {
        initialUserValue = initialValue
    }

    const [user, setUser] = useState(initialUserValue)

    useEffect(() => {
        if(user === null) {
            localStorage.removeItem(localStorageKey)
        } else {
            localStorage.setItem(localStorageKey, JSON.stringify(user))
        }
    }, [localStorageKey, user])

    return [user, setUser]
}

export default useLocalStorage;