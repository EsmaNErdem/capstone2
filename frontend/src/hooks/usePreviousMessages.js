import { useState, useEffect } from "react";
import BookClubApi from "../api";

/** Custom Hook for fetching previous messages between receiver and current user
 * This hook makes API call to backend to retrieve related data and set messaegs state
 * 
 * Returns error and messages
 */
const usePreviousMessages = (receiver, currentUserName, websocket, isOpen) => {
    const [error, setError] = useState(null);
    const [messages, setMessages] = useState([]);

    /**
     * Fetches the previous messages between users from backend sends to child component after WebSocket connection is established
     */ 
    useEffect(function getPreviousMessages() {
        console.debug("usePreviousMessages hook useEffect getPreviousMessages");
    
        const getMessages = async () => {
            try{
                if (isOpen && websocket) {
                    const roomName = `${receiver},${currentUserName}`
                    const previousMessages = await BookClubApi.getRoomPreviousMessages(roomName);
                    setMessages([...previousMessages]);                 
                }
            } catch (e) {
                console.error("usePreviousMessages useEffect API call data loading error:", e);
                setError("An error occurred while fetching previous messages.");
            }
        }
        getMessages();
    }, [isOpen, receiver, websocket]);

    return { error, messages }
}

export default usePreviousMessages;