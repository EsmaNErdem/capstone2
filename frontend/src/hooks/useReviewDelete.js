import { useState, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review delete
 * This hook provides functionality to manage user review delete and updateting bookReviews state
 * 
 * -Utilizes useContext data to delete user review from database
 * -Updates book review
 * 
 * Returns error and addBookReview function
 */
const useReviewDelete = (setBookReviews) => {
    console.debug("useReviewDelete"); 

    const { deleteUserReview } = useContext(UserContext);
    const [error, setError] = useState(null);

    // deletes current user's book review, function is called in BookReview
    const deleteBookReview = async (reviewId, bookId) => {
        try {
            const deletedReviewId = await deleteUserReview(reviewId);
            if(reviewId === +deletedReviewId){
                // Update the state to remove the deleted review
                setBookReviews((prevBookReviews) => {
                    const updatedBookReviews = prevBookReviews.filter(review => review.reviewId !== +deletedReviewId);
                    return updatedBookReviews;
                });
            }
        } catch (error) {
            setError("Error deleting book review.")
            console.error("Error deleting book review:", error);
        }
    } 
    
    return { error, deleteBookReview }

}

export default useReviewDelete