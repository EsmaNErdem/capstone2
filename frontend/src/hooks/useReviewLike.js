import { useState, useEffect, useContext } from 'react';
import UserContext from '../auth/UserContext';

/** Custom Hook for managing user review like
 * This hook provides functionality to manage user likes on reviews, including retrieving initial like status and count,
 * updating the like status and count, and handling user interactions to like or unlike a review.
 * 
 * -Utilizes useContext data to send user like to database
 * -Creates state to update review like count and user review like status
 * -Use useEffect to avoid to recalculations
 * 
 * Returns user like status, review like count states and handleLike function
 */
const useReviewLike = (reviewId, reviewLikeCount) => {   
    console.debug("useReviewLike"); 
  
    const { hasLikedReview, likeReview } = useContext(UserContext);
    const [liked, setLiked] = useState();
    const [likes, setLikes] = useState(reviewLikeCount);
    const [error, setError] = useState(null);

    /**By using the useEffect, the liked status is only recalculated when the id avoiding unnecessary recalculations on every render.  */
    useEffect(() => {
        console.debug("BookReview like useEffect, reviewId=", reviewId)

        setLiked(hasLikedReview(reviewId))
    }, [reviewId])
    
    /**
     * Send user like to database
     * Updates the like count and status based on user actions.
     */ 
    const handleLikeReview = async () =>{
        try {
            setError(null);
            if (liked) {
              const unlikeReviewId = await likeReview(reviewId);

              if (unlikeReviewId === reviewId) {
                setLikes((likes) => likes - 1);
                setLiked(false);
                
              }
            } else {
              const likeReviewId = await likeReview(reviewId);

              if (likeReviewId === reviewId) {
                setLikes((likes) => likes + 1);
                setLiked(true);
              } 
            }
        } catch (error) {
          setError("Error liking review.")
          console.error("Error handling review like:", error);
        }
    }

    return { liked, likes, error, handleLikeReview };
}

export default useReviewLike;