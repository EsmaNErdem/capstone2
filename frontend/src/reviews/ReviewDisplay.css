/* ReviewDisplay.css */
.ReviewDisplay {
  display: flex;
  flex-direction: column;
  align-items: stretch;
  padding: 16px;
  border-bottom: 1px solid #ddd;
}

.UserInfo {
  width: 100%;
  display: flex;
  align-items: stretch;
  justify-content: space-between;

}

.UsernameLink {
  text-decoration: none;
  color: #6d17b7;
  display: flex;
  align-items: center;
}

.UsernameLink:hover {
  text-decoration: underline;
}

.Avatar {
  width: 32px;
  height: 32px;
  margin-right: 8px;
}

.ReviewInfo {
  width: 80%;
  margin: 20px 0px;
}

.BookInfo {
  width: 80%;
  display: flex;
  align-items: stretch;
  justify-content:space-evenly;
  margin: 10px 0px;
}

.TitleLink {
  text-decoration: none;
  color: #a074f3;
}

.TitleLink:hover {
  text-decoration: underline;
}

.CardImage {
  width: 100px;
  height: 150px;
  object-fit: cover;
  margin-right: 30px;
}
