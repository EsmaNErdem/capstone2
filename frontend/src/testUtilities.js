import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  username: "testuser",
  firstNsme: "testfirst",
  lastName: "testlast",
  email: "test@test.net",
  img: null,
};

const UserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => false, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const CurrUserProvider =
    ({ children, currentUser = demoUser, hasLikedReview = () => false, hasLikedBook = () => false, isUserReview = () => true, hasFollowing = () => false, setCurrentUser = u => u }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing, setCurrentUser, isUserReview }}>
      {children}
    </UserContext.Provider>
);

const NonUserProvider =
    ({ children, currentUser = null, hasLikedReview = () => false, hasLikedBook = () => false, hasFollowing = () => false }) => (
    <UserContext.Provider value={{ currentUser, hasLikedReview, hasLikedBook, hasFollowing,}}>
      {children}
    </UserContext.Provider>
);

export { UserProvider, CurrUserProvider, NonUserProvider };
