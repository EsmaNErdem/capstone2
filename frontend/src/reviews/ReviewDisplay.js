import React, { useContext, useState, useEffect } from 'react';
import { Link } from 'react-router-dom';
import UserContext from '../auth/UserContext';
import useReviewLike from '../hooks/useReviewLike';
import Alert from '../utilities/Alert';
import { Divider, ListItem, ListItemText, Avatar, IconButton } from '@mui/material';
import FavoriteIcon from '@mui/icons-material/Favorite';
import DeleteIcon from '@mui/icons-material/Delete';
import './ReviewDisplay.css';

/**
 * Show review with a like button and adds likes to review 
 * if current user review show delete button adn delete review
 * 
 * - Rendered by ReviewList and ReviewFilterList to display reviews and related books
 * - Send user like on a review. If fails, shows error to user
 * 
 * - ReviewList, ReviewFilterList ==> ReviewCard
 */
const ReviewDisplay = ({ reviewId, review, date, username, userImg, reviewLikeCount=0, deleteReview, bookId, title, author=null, cover=null, category=null, likedReview=false, setReviews=null }) => {
  console.debug('ReviewDisplay');

  const { isUserReview } = useContext(UserContext);

  const [userReview, setUserReview] = useState();
  const [error, setError] = useState(null);
  const { liked, likes, error: likeError, handleLikeReview } = useReviewLike(reviewId, reviewLikeCount);

  /**By using the useEffect, the liked status is only recalculated when the id or the   */
  useEffect(() => {
    console.debug('ReviewDisplay useEffect, reviewId=', reviewId);

    setUserReview(isUserReview(reviewId));
  }, [reviewId, isUserReview]);

  // Delete current user review
  const handleDeleteReview = async () => {
    try {
      setError(null);
      await deleteReview(reviewId, bookId);
    } catch (error) {
      setError('Error deleting review.');
      console.error('Error handling review delete:', error);
    }
  };

  const handleLikeReviewButton = async () => {
    await handleLikeReview();
    if(liked && likedReview) {
      setReviews(prevReviews => {
        const updatedReviews = prevReviews.filter(review => review.reviewId !== reviewId)
        return updatedReviews
      })
    }
  }

  return (
    <React.Fragment>
      <ListItem disablePadding className="ReviewDisplay">
        <div className="UserInfo">
          <Link to={`/profile/${username}`} className="UsernameLink" data-testid="review-display-profile-link">
            <Avatar alt={username} src={userImg} sx={{ width: 32, height: 32, marginRight: 2 }} />
            <ListItemText 
              primary={username} 
              secondary={`${new Date(date).toLocaleDateString()} ${new Date(
                date
              ).toLocaleTimeString()}`}
            />
          </Link>
          {userReview ? 
          (
            <>
              <div style={{ display: "flex", alignItems: "stretch" }}>
                <IconButton color="success">
                  <span>{likes}</span>
                  <FavoriteIcon />
                </IconButton>
                <IconButton onClick={handleDeleteReview} color="secondary" data-testid="review-display-review-delete">
                    <DeleteIcon />
                </IconButton>
              </div>
            </>
          ) : (
            <div>
              <IconButton onClick={handleLikeReviewButton} color={liked ? 'error' : 'default'} data-testid="review-display-review-like">
                <span data-testid="display-like-count">{likes}</span>
                <FavoriteIcon />
              </IconButton>
            </div>
          )}
        </div>
        <div className="ReviewInfo">
          <ListItemText
            primaryTypographyProps={{fontSize: '18px', fontWeight: "600"}} 
            primary={review}
          />
        </div>
        {title && (
          <div className="BookInfo">
            <Link to={`/books/${bookId}`} className="Title">
                <img alt={title} src={cover} className="CardImage"/>
            </Link>
            <div>
                <h2>
                <Link to={`/books/${bookId}`} className="TitleLink" data-testid="review-display-book-title-link">
                    {title}
                </Link>
                </h2>
                {author && <p>Author: {author}</p>}
                {category && <p>Category: {category}</p>}
            </div>
          </div>
        )}
      </ListItem>
      {likeError && <Alert type="danger" messages={[likeError]} />}
      {error ? <Alert type="danger" messages={[error]} /> : null}
      <Divider />
    </React.Fragment>
  );
};

export default ReviewDisplay;
