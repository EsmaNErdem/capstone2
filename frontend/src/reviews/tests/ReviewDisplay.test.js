import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { UserProvider, CurrUserProvider } from "../../testUtilities";
import { render, fireEvent, act, waitFor } from '@testing-library/react';
import ReviewDisplay from '../ReviewDisplay';
import * as useReviewLikeModule from '../../hooks/useReviewLike';
import * as useReviewDeleteModule from '../../hooks/useReviewDelete';

const review = {
    reviewId: 1,
    review: "Review 1",
    date: "12/18/2023",
    username: "Test1", 
    userImg: "UserImg1",
    book_id: "1",
    title: 'Book 1',
    cover: 'Cover 1',
    author: 'Author 1',
    description: 'Description 1',
    category: 'Category 1',
    reviewLikeCount: 2

}

jest.mock("../../api", () => ({
    __esModule: true,
    default: {
        likeReview: jest.fn(),
        deleteBookReview: jest.fn(),
    },
}));

let handleLikeReviewMock;
let handleDeleteReviewMock;

beforeEach(() => {
  handleLikeReviewMock = jest.fn();
  handleDeleteReviewMock = jest.fn();

  jest.spyOn(useReviewLikeModule, 'default').mockImplementation(() => ({
    liked: false,
    likes: 2,
    error: null,
    handleLikeReview: handleLikeReviewMock,
  }));

  jest.spyOn(useReviewDeleteModule, 'default').mockImplementation(() => ({
    error: null,
    handleDeleteReviewMock: handleDeleteReviewMock,
  }));
});
  
afterEach(() => {
    jest.clearAllMocks();
});

test('renders without crashing', async () => {
    await act(async () => {
        render(
            <MemoryRouter>
                <UserProvider>
                    <ReviewDisplay />
                </UserProvider>
            </MemoryRouter>
        );
    });

});

test('it renders and matches with snapshot', async () => {
    await act(async () => {
            ({ asFragment } = render(
                <MemoryRouter>
                    <UserProvider>
                        <ReviewDisplay
                              reviewId={review.id}
                              review={review.review}
                              date={review.date}
                              username={review.username} 
                              userImg={review.userImg}
                              bookId={review.book_id}
                              title={review.title}
                              author={review.author}
                              cover={review.cover}
                              category={review.category}
                              reviewLikeCount={review.reviewLikeCount}
                        />
                    </UserProvider>
                </MemoryRouter>
            ));
        });
        
    expect(asFragment()).toMatchSnapshot();

    await act(async () => {
        ({ asFragment } = render(
            <MemoryRouter>
                <CurrUserProvider>
                    <ReviewDisplay
                          reviewId={review.id}
                          review={review.review}
                          date={review.date}
                          username={review.username} 
                          userImg={review.userImg}
                          bookId={review.book_id}
                          title={review.title}
                          author={review.author}
                          cover={review.cover}
                          category={review.category}
                          reviewLikeCount={review.reviewLikeCount}
                    />
                </CurrUserProvider>
            </MemoryRouter>
        ));
    });
    
expect(asFragment()).toMatchSnapshot();
});

test('render review and links to the correct book details page', async () => {
    await act(async () => {
        ({ getByTestId, getByText } = render(
            <MemoryRouter>
                <UserProvider>
                    <ReviewDisplay
                        reviewId={review.id}
                        review={review.review}
                        date={review.date}
                        username={review.username} 
                        userImg={review.userImg}
                        bookId={review.book_id}
                        title={review.title}
                        author={review.author}
                        cover={review.cover}
                        category={review.category}
                        reviewLikeCount={review.reviewLikeCount}
                    />
                </UserProvider>
            </MemoryRouter>
        ));
    });
    
    expect(getByText("Review 1")).toBeInTheDocument()
    expect(getByTestId("display-like-count")).toHaveTextContent("2")
    const titleLink = getByTestId('review-display-book-title-link');
    expect(titleLink.getAttribute('href')).toBe('/books/1');
    const profileLink = getByTestId('review-display-profile-link');
    expect(profileLink.getAttribute('href')).toBe('/profile/Test1');
});

test('implement review like functionality', async () => {
    await act(async () => {
        ({ getByTestId, getByText } = render(
            <MemoryRouter>
                <UserProvider>
                    <ReviewDisplay
                        reviewId={review.id}
                        review={review.review}
                        date={review.date}
                        username={review.username} 
                        userImg={review.userImg}
                        bookId={review.book_id}
                        title={review.title}
                        author={review.author}
                        cover={review.cover}
                        category={review.category}
                        reviewLikeCount={review.reviewLikeCount}
                    />
                </UserProvider>
            </MemoryRouter>
        ));
    });
    
    const reviewLikeButton = getByTestId('review-display-review-like');
    fireEvent.click(reviewLikeButton);
    
    await waitFor(() => {
        expect(handleLikeReviewMock).toHaveBeenCalled();
    });
});

test('implement review like functionality', async () => {
    await act(async () => {
        ({ getByTestId, getByText } = render(
            <MemoryRouter>
                <CurrUserProvider>
                    <ReviewDisplay
                        reviewId={review.id}
                        review={review.review}
                        date={review.date}
                        username={review.username} 
                        userImg={review.userImg}
                        bookId={review.book_id}
                        title={review.title}
                        author={review.author}
                        cover={review.cover}
                        category={review.category}
                        reviewLikeCount={review.reviewLikeCount}
                    />
                </CurrUserProvider>
            </MemoryRouter>
        ));
    });
    
    const reviewDeleteButton = getByTestId('review-display-review-delete');
    fireEvent.click(reviewDeleteButton);
    
    await waitFor(() => {
        expect(getByText("Error deleting review.")).toBeInTheDocument()
    });
});