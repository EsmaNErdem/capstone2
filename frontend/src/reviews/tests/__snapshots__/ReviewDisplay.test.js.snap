// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders and matches with snapshot 1`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters ReviewDisplay css-1nxmd3h-MuiListItem-root"
  >
    <div
      class="UserInfo"
    >
      <a
        class="UsernameLink"
        data-testid="review-display-profile-link"
        href="/profile/Test1"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
        >
          <img
            alt="Test1"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="UserImg1"
          />
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Test1
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
          >
            12/18/2023 12:00:00 AM
          </p>
        </div>
      </a>
      <div>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
          data-testid="review-display-review-like"
          tabindex="0"
          type="button"
        >
          <span
            data-testid="display-like-count"
          >
            2
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="ReviewInfo"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1k48cxs-MuiTypography-root"
        >
          Review 1
        </span>
      </div>
    </div>
    <div
      class="BookInfo"
    >
      <a
        class="Title"
        href="/books/1"
      >
        <img
          alt="Book 1"
          class="CardImage"
          src="Cover 1"
        />
      </a>
      <div>
        <h2>
          <a
            class="TitleLink"
            data-testid="review-display-book-title-link"
            href="/books/1"
          >
            Book 1
          </a>
        </h2>
        <p>
          Author: Author 1
        </p>
        <p>
          Category: Category 1
        </p>
      </div>
    </div>
  </li>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
  />
</DocumentFragment>
`;

exports[`it renders and matches with snapshot 2`] = `
<DocumentFragment>
  <li
    class="MuiListItem-root MuiListItem-gutters ReviewDisplay css-1nxmd3h-MuiListItem-root"
  >
    <div
      class="UserInfo"
    >
      <a
        class="UsernameLink"
        data-testid="review-display-profile-link"
        href="/profile/Test1"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular css-1fy7wd8-MuiAvatar-root"
        >
          <img
            alt="Test1"
            class="MuiAvatar-img css-1pqm26d-MuiAvatar-img"
            src="UserImg1"
          />
        </div>
        <div
          class="MuiListItemText-root MuiListItemText-multiline css-konndc-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
          >
            Test1
          </span>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiListItemText-secondary css-83ijpv-MuiTypography-root"
          >
            12/18/2023 12:00:00 AM
          </p>
        </div>
      </a>
      <div
        style="display: flex; align-items: stretch;"
      >
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSuccess MuiIconButton-sizeMedium css-1r6h6t4-MuiButtonBase-root-MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span>
            2
          </span>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="FavoriteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSecondary MuiIconButton-sizeMedium css-1gws2xf-MuiButtonBase-root-MuiIconButton-root"
          data-testid="review-display-review-delete"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="DeleteIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      class="ReviewInfo"
    >
      <div
        class="MuiListItemText-root css-tlelie-MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-1k48cxs-MuiTypography-root"
        >
          Review 1
        </span>
      </div>
    </div>
    <div
      class="BookInfo"
    >
      <a
        class="Title"
        href="/books/1"
      >
        <img
          alt="Book 1"
          class="CardImage"
          src="Cover 1"
        />
      </a>
      <div>
        <h2>
          <a
            class="TitleLink"
            data-testid="review-display-book-title-link"
            href="/books/1"
          >
            Book 1
          </a>
        </h2>
        <p>
          Author: Author 1
        </p>
        <p>
          Category: Category 1
        </p>
      </div>
    </div>
  </li>
  <hr
    class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
  />
</DocumentFragment>
`;
